CORE-12871: Avoid softlocking adjacent segment merger by skipping across compacted segments

oleiman opened the pull request:
Primarily this:

Adjacent segment merging only runs when compaction is disabled, but if
compaction was previous enabled for some topic then there may be compacted
segments on disk for a while (or indefinitely depending on the retention
policy).

If these segments:
- Were uploaded to cloud storage before compaction
- Were NOT reuploaded after compaction
- Were originally small enough to be mergeable relative to the configured
  remote segment target size.

Then the adjacent segment merger can get in a nasty situation where it scans
the partition manifest for a run of small segments and proposes that run for
merging but the sizes recorded in the partition manifest don't match the
actual (compacted) size of the segments on disk. Segment collection can
succeed but the content size of the resulting upload candidate doesn't match
the run and so is rejected by the housekeeping job.

Since no upload took place, the '_last' offset pointer in the housekeeping job
is not updated, so the merger attempts to reupload the same, incorrectly sized
run effectively forever (or until the offending compacted segments are removed
from the local log).

This commit introduces 'find_reupload_candidate_result::skip_to', an optional
rp offset that indicates to the housekeeping job that, although no upload
took place, the '_last' pointer should still be updated so that the merger
can make progress.

Also fixes a bug where adjacent_segment_merger would usually ignore the advancing last uploaded offset, scanning from the remote log on each run.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Bug Fixes

* Fixes a minor bug in cloud storage housekeeping where adjacent segment merge jobs could fail to make progress after compaction was switched off.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
archival: adjacent_segment_merger should not get stuck on a bad run

Adjacent segment merging only runs when compaction is disabled, but if
compaction was previous enabled for some topic then there may be compacted
segments on disk for a while (or indefinitely depending on the retention
policy).

If these segments:
- Were uploaded to cloud storage before compaction
- Were NOT reuploaded after compaction
- Were originally small enough to be mergeable relative to the configured
  remote segment target size.

Then the adjacent segment merger can get in a nasty situation where it scans
the partition manifest for a run of small segments and proposes that run for
merging but the sizes recorded in the partition manifest don't match the
actual (compacted) size of the segments on disk. Segment collection can
succeed but the content size of the resulting upload candidate doesn't match
the run and so is rejected by the housekeeping job.

Since no upload took place, the '_last' offset pointer in the housekeeping job
is not updated, so the merger attempts to reupload the same, incorrectly sized
run effectively forever (or until the offending compacted segments are removed
from the local log).

This commit introduces 'find_reupload_candidate_result::skip_to', an optional
rp offset that indicates to the housekeeping job that, although no upload
took place, the '_last' pointer should still be updated so that the merger
can make progress.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
archival: adjacent_segment_merger jobs should scan from the last upload

Fixes a bug where in all runs but the very first, the adjacent segment merger
job would override the '_last' offset (end offset of the last successful upload)
with the manifest start offset, mistakenly treating every run like a remote
segment merge run, which we never do anyway.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dt/adjacent_merge: Test for compaction toggle edge case

Checks that adjacent segment merging can make progress even when the fron of
the local log is already compacted.

oleiman committed:
dt: Refactor tests/adjacent_segment_merging_test.py

Base class for reuse

WillemKauf commented:
> Then the adjacent segment merger can get in a nasty situation where it scans
the partition manifest for a run of small segments and proposes that run for
merging but the sizes recorded in the partition manifest don't match the
actual (compacted) size of the segments on disk. Segment collection can
succeed but the content size of the resulting upload candidate doesn't match
the run and so is rejected by the housekeeping job.

This nasty situation is a softlock of future merger runs, right?

WillemKauf commented on src/v/cluster/archival/adjacent_segment_merger.cc:
```diff
@@ -89,7 +89,7 @@ std::optional<adjacent_segment_run> adjacent_segment_merger::scan_manifest(
           manifest.get_start_offset().value_or(local_start_offset),
           local_start_offset);
-    } else {
+    } else if (!_is_local) {
```
I only see one relevant construction site of the `adjacent_segment_merger` in the code base where `_is_local` is passed as `true`.

https://github.com/redpanda-data/redpanda/blob/faf98950bd41a340f78bad2627ac50bc21d766e3/src/v/cluster/archival/ntp_archiver_service.cc#L274-L280 

Is any use of this `_is_local` variable essentially dead code?

oleiman replied:
well, the lore goes that we have a remote segment merge feature that was built but never turned on. remove or enable is different, larger task i think.

oleiman replied:
> 😕 

indeed

Lazin replied:
It was never merged because there were performance concerns (you need to upload 1000s of small segments to meet segment size requirements).

WillemKauf commented on src/v/cluster/archival/ntp_archiver_service.h:
```diff
@@ -342,6 +342,7 @@ class ntp_archiver {
         std::optional<ssx::checkpoint_mutex_units> units;
         std::optional<segment_collector_stream> upload_stream{};
         archival_stm_fence read_write_fence{};
+        std::optional<model::offset> skip_to{};
```
Please add a comment describing this variable's purpose in the code here. There are a lot of variables  in `archival` that I _wish_ had comments describing them. Instead, the reader has to grok commit messages/CI failure history, read code paths, etc...

Does this variable _only_ target this specific edge case with compaction?

oleiman replied:
> Does this variable only target this specific edge case

yeah, as written. i think there's at least one other situation where we might want to use it, but i need to think about it a bit more. this one is straightforwardly a bug.

Lazin replied:
great comment

WillemKauf commented on tests/rptest/tests/adjacent_segment_merging_test.py:
```diff
@@ -125,3 +125,111 @@ def manifest_has_one_segment():
+        self.logger.debug(
+            "The log is full of small compacted segments, so housekeeping shouldn't have any effect"
+        )
+        with expect_exception(TimeoutError, lambda _: True):
```
Nice test. Would be cool to see `expect_timeout()` pulled out as an ducktape abstraction in the future :)

oleiman replied:
ya I've been meaning to do that

oleiman replied:
done

oleiman committed:
dt/util: Introduce expect_timeout

Thin convenience wrapper around expect_exception.

- klass: 'ducktape.errors.TimeoutError'
- validator: 'lambda _: True'

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--42363fa1-010a-407b-94d5-5ebcaf8769a8-->
<!--TEST RESULTS-->
<!--3f57d0c5-ed7d-4f4c-8718-d10ff0cb66c6-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71655">build#71655</a></summary>


| test_class                              | test_method                             | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------------------|:----------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest                | test_node_operations                    | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71655#0198f8dc-b70f-4c7a-b9c6-a42f30d48c95 | FLAKY         | 19/21    | upstream reliability is '99.13419913419914'. current run reliability is '90.47619047619048'. drift is 8.65801 and the allowed drift is set to 50. The test should PASS |
| RapidTopicRecreateTest                  | test_topic_rapid_recreation             | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/71655#0198f8dc-b713-4646-a2d7-4369584a5da6 | FLAKY         | 19/21    | upstream reliability is '99.48849104859335'. current run reliability is '90.47619047619048'. drift is 9.0123 and the allowed drift is set to 50. The test should PASS  |
| src/v/cluster/tests/tx_compaction_tests | src/v/cluster/tests/tx_compaction_tests |                                                                                                                                         | unit        | https://buildkite.com/redpanda/redpanda/builds/71655#0198f8aa-a73d-40b2-a8d7-eb569cc9983f | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71659">build#71659</a></summary>


| test_class            | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71659#0198fca2-d842-487e-8e03-56ecf662c71c | FLAKY         | 20/21    | upstream reliability is '99.21414538310412'. current run reliability is '95.23809523809523'. drift is 3.97605 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

oleiman commented:
> This nasty situation is a softlock of future merger runs, right?

Yup. ~~Cleverly left out of the commit message... will add.~~ nvm next paragraph

WillemKauf reviewed (approved):
nice

Lazin commented on src/v/cluster/archival/ntp_archiver_service.cc:
```diff
@@ -3496,6 +3495,15 @@ ntp_archiver::find_reupload_candidate(
+                    "Failed to make reupload candidate due to size mismatch, "
+                    "skip this range: expected size: {}, actual size: {}",
+                    human::bytes(run->meta.size_bytes),
+                    human::bytes(collector_stream.size));
```
can you also include size in bytes (sometimes it's useful to grep the logs)?

oleiman replied:
you mean the text "size in bytes"? or did i miss a value?

Lazin reviewed (approved):
the code LGTM with one minor nit

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-27440-v25.2.x-130 remotes/upstream/v25.2.x
git cherry-pick -x e359b0373e c994a11a36 a0ad9f4b5e fd419cbbaa d045933af3
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/17510123511).
